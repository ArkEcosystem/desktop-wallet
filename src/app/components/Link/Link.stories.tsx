import React from "react";

import { Link } from "./Link";

export default { title: "App / Components / Link" };

export const Default = () => <Link to="/wallets">Wallets</Link>;
export const External = () => (
	<Link to={{ pathname: "https://ark.io" }} isExternal>
		ARK.io
	</Link>
);
export const WithTooltip = () => (
	<Link to={{ pathname: "https://ark.io" }} isExternal tooltip="ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT" />
);
