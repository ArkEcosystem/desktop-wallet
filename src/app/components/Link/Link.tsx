import { Tooltip } from "app/components/Tooltip";
import { toasts } from "app/services";
import React from "react";
import { useTranslation } from "react-i18next";
import { Link as RouterLink, LinkProps } from "react-router-dom";
import tw, { css, styled } from "twin.macro";
import { openExternal } from "utils/electron-utils";

import { Icon } from "../Icon";

const AnchorStyled = styled.a<{ isExternal: boolean }>`
	${({ isExternal }) =>
		isExternal &&
		css`
			&:hover,
			&:active {
				${tw`no-underline`}

				.underline-dotted {
					${tw`relative`}

					&:after {
						content: "";
						${tw`block w-full border-b absolute bottom-0 border-dotted`}
					}
				}
			}
		`}
`;

type AnchorProps = {
	isExternal?: boolean;
	navigate?: () => void;
	showExternalIcon?: boolean;
} & React.AnchorHTMLAttributes<any>;

const Anchor = React.forwardRef<HTMLAnchorElement, Props>(
	({ isExternal, showExternalIcon, children, rel, ...props }: AnchorProps, ref) => (
		<AnchorStyled
			data-testid="Link"
			isExternal={isExternal!}
			className="inline-flex items-center font-semibold transition-colors duration-200 cursor-pointer text-theme-primary-600 hover:text-theme-primary-700 hover:underline active:text-theme-primary-500"
			rel={isExternal ? "noopener noreferrer" : rel}
			ref={ref}
			onClick={(event) => {
				event.stopPropagation();
				event.preventDefault();
				return props.navigate?.();
			}}
			{...props}
		>
			<span className="underline-dotted">{children}</span>
			{isExternal && showExternalIcon && (
				<Icon
					data-testid="Link__external"
					name="Redirect"
					className={`flex-shrink-0 ${children ? "ml-2 text-sm" : ""}`}
				/>
			)}
		</AnchorStyled>
	),
);

Anchor.displayName = "Anchor";

type Props = {
	isExternal?: boolean;
	children?: React.ReactNode;
	tooltip?: string;
	showExternalIcon?: boolean;
} & LinkProps;

export const Link = ({ tooltip, ...props }: Props) => {
	const { t } = useTranslation();

	return (
		<Tooltip content={tooltip} disabled={!tooltip}>
			{props.isExternal ? (
				<Anchor
					onClick={(event) => {
						event.stopPropagation();
						event.preventDefault();
						try {
							openExternal(props.to);
						} catch {
							toasts.error(t("COMMON.ERRORS.INVALID_URL", { url: props.to }));
						}
					}}
					{...props}
				/>
			) : (
				<RouterLink component={Anchor} {...props} />
			)}
		</Tooltip>
	);
};

Link.defaultProps = {
	isExternal: false,
	showExternalIcon: true,
};
