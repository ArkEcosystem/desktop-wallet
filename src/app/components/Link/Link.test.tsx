import { act, fireEvent, render } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";

import { Link } from "./Link";

describe("Link", () => {
	it("should render", () => {
		const { getByTestId, asFragment } = render(
			<MemoryRouter>
				<Link to="/test">Test</Link>
			</MemoryRouter>,
		);
		expect(getByTestId("Link")).toHaveTextContent("Test");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render external", () => {
		const { getByTestId, asFragment } = render(
			<MemoryRouter>
				<Link to={{ pathname: "https://ark.io" }} isExternal>
					ARK.io
				</Link>
			</MemoryRouter>,
		);
		expect(getByTestId("Link")).toHaveAttribute("rel", "noopener noreferrer");
		expect(getByTestId("Link")).toHaveAttribute("target", "_blank");
		expect(getByTestId("Link__external")).toBeTruthy();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render external without children", () => {
		const { getByTestId, asFragment } = render(
			<MemoryRouter>
				<Link to={{ pathname: "https://ark.io" }} isExternal />
			</MemoryRouter>,
		);
		expect(getByTestId("Link__external")).toBeTruthy();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with tooltip", () => {
		const { getByTestId, baseElement } = render(
			<MemoryRouter>
				<Link to="/test" tooltip="Custom Tooltip">
					Test
				</Link>
			</MemoryRouter>,
		);

		act(() => {
			fireEvent.mouseEnter(getByTestId("Link"));
		});

		expect(baseElement).toHaveTextContent("Custom Tooltip");
	});
});
