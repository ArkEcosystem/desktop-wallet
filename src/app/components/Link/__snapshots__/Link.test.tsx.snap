// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should open an external link 1`] = `
<DocumentFragment>
  <a
    class="Link__AnchorStyled-sc-1q7sbxr-0 cIuuRE inline-flex items-center font-semibold transition-colors duration-200 cursor-pointer text-theme-primary-600 hover:text-theme-primary-700 hover:underline active:text-theme-primary-500"
    data-testid="Link"
    rel="noopener noreferrer"
    to="https://ark.io/"
  >
    <span
      class="underline-dotted"
    />
    <div
      class="sc-bdfBwQ fPrsah flex-shrink-0 "
      data-testid="Link__external"
      height="1em"
      width="1em"
    >
      <svg>
        redirect.svg
      </svg>
    </div>
  </a>
</DocumentFragment>
`;

exports[`Link should render 1`] = `
<DocumentFragment>
  <a
    class="Link__AnchorStyled-sc-1q7sbxr-0 inline-flex items-center font-semibold transition-colors duration-200 cursor-pointer text-theme-primary-600 hover:text-theme-primary-700 hover:underline active:text-theme-primary-500"
    data-testid="Link"
    href="/test"
  >
    <span
      class="underline-dotted"
    >
      Test
    </span>
  </a>
</DocumentFragment>
`;

exports[`Link should render external without children 1`] = `
<DocumentFragment>
  <a
    class="Link__AnchorStyled-sc-1q7sbxr-0 cIuuRE inline-flex items-center font-semibold transition-colors duration-200 cursor-pointer text-theme-primary-600 hover:text-theme-primary-700 hover:underline active:text-theme-primary-500"
    data-testid="Link"
    rel="noopener noreferrer"
    to="https://ark.io"
  >
    <span
      class="underline-dotted"
    />
    <div
      class="sc-bdfBwQ fPrsah flex-shrink-0 "
      data-testid="Link__external"
      height="1em"
      width="1em"
    >
      <svg>
        redirect.svg
      </svg>
    </div>
  </a>
</DocumentFragment>
`;

exports[`Link should render with tooltip 1`] = `
<DocumentFragment>
  <a
    class="Link__AnchorStyled-sc-1q7sbxr-0 inline-flex items-center font-semibold transition-colors duration-200 cursor-pointer text-theme-primary-600 hover:text-theme-primary-700 hover:underline active:text-theme-primary-500"
    data-testid="Link"
    href="/test"
  >
    <span
      class="underline-dotted"
    >
      Test
    </span>
  </a>
</DocumentFragment>
`;

exports[`Link should show a toast when trying to open an invalid external link 1`] = `
<DocumentFragment>
  <a
    class="Link__AnchorStyled-sc-1q7sbxr-0 cIuuRE inline-flex items-center font-semibold transition-colors duration-200 cursor-pointer text-theme-primary-600 hover:text-theme-primary-700 hover:underline active:text-theme-primary-500"
    data-testid="Link"
    rel="noopener noreferrer"
    to="invalid-url"
  >
    <span
      class="underline-dotted"
    />
    <div
      class="sc-bdfBwQ fPrsah flex-shrink-0 "
      data-testid="Link__external"
      height="1em"
      width="1em"
    >
      <svg>
        redirect.svg
      </svg>
    </div>
  </a>
</DocumentFragment>
`;
